### ISSUE TITLE
Update profile search filters to determine criteria type

### ISSUE THREAD

**Issue by pjmonks** [MauroDataMapper/mdm-core#399]
# Description

Part of MauroDataMapper/mdm-ui#774 requires some adjustments to the existing `ProfileFilter` class in `mdm-plugin-profile`. Update this `SearchParamFilter` to return different types of criteria for wider support.

# Current implementation

To use the `ProfileFilter`, a search request must contain a JSON body similar to below:

```json
{
  "profileFields": [
    {
      "metadataNamespace": "...",
      "metadataPropertyName": "...",
      "filterTerm": "..."
    }
  ]
}
```

The `ProfileFilter` will identify that a `profileFields` array exists in the search params map and iterate over the array creating a `phrase` criteria for this metadata key.

This has some disadvantages:

1. Using `phrase` means that the metadata namespace, key and value must _exactly_ match an existing value in the metadata table.
2. These inputs are lengthy - namespaces are long strings, keys could be anything, and values are ultimately free text strings. Matching these exactly will prove to be cumbersome.

# Suggested implementation

Deciding the correct namespace, key and value to use will be part of MauroDataMapper/mdm-ui#774. For value search though, it is suggested that `ProfileFilter` be updated to accept a new parameter that can choose the search criteria type. For instance `phrase` is useful for an exact match on a known value (such as an enumeration type), but `keyword` is more useful for free text input.

The JSON body could potentially include a parameter like this:

```json
{
  "profileFields": [
    {
      "metadataNamespace": "...",
      "metadataPropertyName": "...",
      "filterTerm": "...",
      "type": "phrase"  // phrase | keyword (default: phrase)
    }
  ]
}
```

This will provide more control over the search over all metadata.

**Comment by joe-crawford** [comment]
Hi @pjmonks, I'm testing the existing ProfileFilter endpoint, when a model is present with this metadata:
```json
{
    "id": "77983ab9-32ea-4138-932c-619f8cf50f6d",
    "namespace": "ox.ac.uk",
    "key": "key1",
    "value": "Soluta expedita ad qui et sequi optio ratione omnis.",
    "lastUpdated": "2023-03-06T13:49:50.01Z"
}
```
the phrase search does match on words within the value:
```json
{
    "profileFields": [
        {"metadataNamespace": "ox.ac.uk", "metadataPropertyName": "key1", "filterTerm": "et"}
    ]
}
```
returns
```
{
    "count": 1,
    "items": [
        {
            "id": "85777cf2-d5fa-4b56-86fc-f4136a3c05cd",
            "domainType": "DataModel",
            "label": "search testing model 1"
        }
    ]
}
```


So I think the `phrase` type is actually the right one for free text search? Hibernate says it "matches documents for which a given field contains a given sequence of words, in the given order".

Do we still want to add an exact match option for the Enumeration Value use case?

(I'm looking at the list of Hibernate predicates in: https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-predicate)

**Comment by pjmonks** [comment]
@joe-crawford I might not be understanding the match options available in Hibernate then. That example you gave works certainly, then I tried a similar one on my database:

```json
{
  "metadataNamespace": "uk.ac.ox.softeng.maurodatamapper.plugins.explorer.research",
  "metadataPropertyName": "databaseName",
  "filterTerm": "laboratory_tests"
}
```

Results in expected data elements returned. However, if I change the `filterTerm` to be a partial value, like "lab", then zero results are returned. This particular profile field would be represented as text instead of enumeration, so I feel like it is not unreasonable for a user to type part of a word and expect some matches back. The `phrase` does seem to do this - if a metadata value is composed of a sentence, with words and spaces, then it seems to match only if you've typed the exact _word_.

In comparison, doing a simple search term like this:

```json
{
  "searchTerm": "batch",  
}
```

Would give results like this:

```json
{
  "count": 10,
  "items": [
    {
      "id": "2a182cb8-18d1-42c5-b1b5-9454de5d7795",
      "domainType": "DataElement",
      "label": "Batch test code",
      "model": "7902f836-7509-4237-8efc-50e098983322",
    },
    {
      "id": "ebc9eedd-b614-456d-83fd-7e92db5bc31a",
      "domainType": "DataElement",
      "label": "CDS_BATCH_CONTENT_ALIAS_UPDT_DT_TM",
      "model": "1d2843ec-afd8-44a5-a7e1-7823dc12faba",
    },
  ]
}
```

In this case the word "batch" matches both upper/lowercase and as part of a sentence or part of a larger word. Should a metadata value be attempting to match in this way too?

**Comment by joe-crawford** [comment]
@pjmonks, you're right, I will add the keyword matching as intended, I was wondering why the `phrase` seemed to match on keywords as well but I agree the partial matching would be useful, and I think the phrase matching the words only in the given order makes it suitable for enumeration value keys as well.

**Comment by joe-crawford** [comment]
@pjmonks, I've added some examples in the pull request, but I have added the search type that I think is wanted as `"type": "contains"`.

In the `batch` example above, I think the search is finding the second element (CDS_BATCH_CONTENT_ALIAS_UPDT_DT_TM) because the labels are being split on underscores when being tokenised, but the metadata values are only split on spaces. I think this is OK as I can imagine wanting to search for a metadata value 'FOO_BAR' separately from 'BAR_FOO'.

I don't think the second element would be found above if doing a simple search on `bat` because the `simpleSearchQuery` type (used by CatalogueItem label search) doesn't search for text contained within tokens. I've added a `simpleSearchQuery` to ProfileFilter as `"type": "query"` but that doesn't change the tokeniser for the metadata values.
